Eliminate fatal to-dos from C# parser

[ERROR] Semgrep Core — Fatal errorAn error occurred while invoking the Semgrep engine. Please help us fix this by creating an issue at https://github.com/returntocorp/semgrepAt line src/WinSW.Core/Download.cs:1: Common.Impossible====[ BEGIN error trace ]====Raised at Parse_csharp_tree_sitter.type_ in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2211, characters 6-22Called from Stdlib__option.map in file "option.ml", line 24, characters 57-62Called from Parse_csharp_tree_sitter.explicit_parameter in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2155, characters 11-46Called from Parse_csharp_tree_sitter.formal_parameter_list in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 1960, characters 11-43Called from Parse_csharp_tree_sitter.parameter_list in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2285, characters 16-43Called from Parse_csharp_tree_sitter.statement in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 1752, characters 15-36Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23Called from Parse_csharp_tree_sitter.block in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 807, characters 11-38Called from Parse_csharp_tree_sitter.statement in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 1799, characters 15-27Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Parse_csharp_tree_sitter.block in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 807, characters 11-38Called from Parse_csharp_tree_sitter.function_body in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2127, characters 23-36Called from Parse_csharp_tree_sitter.declaration in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2875, characters 15-35Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Stdlib__list.map in file "list.ml", line 92, characters 32-39Called from Parse_csharp_tree_sitter.declaration_list in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2449, characters 11-42Called from Parse_csharp_tree_sitter.class_interface_struct in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2586, characters 35-58Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23Called from Parse_csharp_tree_sitter.declaration_list in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2449, characters 11-42Called from Parse_csharp_tree_sitter.namespace_declaration in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2549, characters 39-62Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23Called from Parse_csharp_tree_sitter.compilation_unit in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 2535, characters 15-64Called from Parse_csharp_tree_sitter.parse.(fun) in file "src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml", line 3007, characters 12-36Called from Parse_tree_sitter_helpers.wrap_parser in file "src/parsing/tree_sitter/Parse_tree_sitter_helpers.ml", line 155, characters 13-29Called from Parse_target.run_parser in file "src/parsing/Parse_target.ml", line 100, characters 18-24Re-raised at Parse_target.run in file "src/parsing/Parse_target.ml", line 193, characters 26-65Called from Parse_target.parse_and_resolve_name_use_pfff_or_treesitter in file "src/parsing/Parse_target.ml", line 359, characters 30-60Called from Parse_with_caching.parse_generic.(fun) in file "src/runner/Parse_with_caching.ml", line 105, characters 12-80Called from Parse_with_caching.cache_computation.(fun) in file "src/runner/Parse_with_caching.ml", line 58, characters 20-24Called from Parse_with_caching.parse_generic in file "src/runner/Parse_with_caching.ml", line 89, characters 4-1023Called from CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 31, characters 17-27Re-raised at CamlinternalLazy.force_lazy_block in file "camlinternalLazy.ml", line 36, characters 4-11Called from Common.with_time in file "src/pfff/commons/Common.ml", line 215, characters 12-16Called from Match_search_rules.matches_of_patterns in file "src/engine/Match_search_rules.ml", line 264, characters 8-67Called from Match_search_rules.matches_of_xpatterns in file "src/engine/Match_search_rules.ml", line 601, characters 6-55Called from Match_search_rules.matches_of_formula in file "src/engine/Match_search_rules.ml", line 931, characters 12-63Called from Match_search_rules.check_rule in file "src/engine/Match_search_rules.ml", line 961, characters 4-74Called from Stdlib__list.map in file "list.ml", line 92, characters 20-23Called from Match_search_rules.check in file "src/engine/Match_search_rules.ml", line 988, characters 2-275Called from Match_rules.check in file "src/engine/Match_rules.ml", line 63, characters 4-88Called from Run_semgrep.semgrep_with_rules.(fun) in file "src/runner/Run_semgrep.ml", line 460, characters 13-182Called from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file "src/runner/Run_semgrep.ml", line 305, characters 25-33Called from Common.set_timeout in file "src/pfff/commons/Common.ml", line 1420, characters 12-16Re-raised at Common.set_timeout in file "src/pfff/commons/Common.ml", line 1437, characters 6-13Called from Run_semgrep.timeout_function in file "src/runner/Run_semgrep.ml", line 148, characters 4-73Called from Stdlib__fun.protect in file "fun.ml", line 33, characters 8-15Re-raised at Stdlib__fun.protect in file "fun.ml", line 38, characters 6-52Called from Memory_limit.run_with_memory_limit in file "src/system/Memory_limit.ml", line 79, characters 6-62Called from Run_semgrep.iter_targets_and_get_matches_and_exn_to_errors.(fun) in file "src/runner/Run_semgrep.ml", line 302, characters 17-846=====[ END error trace ]=====

--- a/changelog.d/gh-5068.fixed
+++ b/changelog.d/gh-5068.fixed
@@ -1,0 +1,17 @@
+What:
+Currently, `Impossible` is raised for certain C# programs, when an explicit type is expected 
+where an implicit one was provided (for instance, when introduced by the `var` keyword). An example
+program which triggers this is provided below:
+```
+public class Foo {
+    public void Foo(var a) {
+    }
+}
+```
+
+Why:
+We should reject this program, but with a better error message.
+
+How:
+Change it to be raising an exception, specifically a `Parse_info.Other_error`. This is not a `Parse_info.Parse_error`
+because this is not exactly a parsing bug, c.f. the GitHub discussion in the issue.

--- a/changelog.d/gh-5068.fixed
+++ b/changelog.d/gh-5068.fixed
@@ -1,17 +1,1 @@
-What:
-Currently, `Impossible` is raised for certain C# programs, when an explicit type is expected 
-where an implicit one was provided (for instance, when introduced by the `var` keyword). An example
-program which triggers this is provided below:
-```
-public class Foo {
-    public void Foo(var a) {
-    }
-}
-```
-
-Why:
-We should reject this program, but with a better error message.
-
-How:
-Change it to be raising an exception, specifically a `Parse_info.Other_error`. This is not a `Parse_info.Parse_error`
-because this is not exactly a parsing bug, c.f. the GitHub discussion in the issue.
+C#: Improved error message when function parameters are declared with `var`

--- a/semgrep-core/src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml
+++ b/semgrep-core/src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml
@@ -2308,7 +2308,7 @@ and type_ (env : env) (x : CST.type_) : G.type_ =
   | `Impl_type _tok ->
       (* When type_ is called, we expect an explicit type, not "var".
          The implicit type is handled in local_variable_type. *)
-      raise Impossible
+      raise (Parse_info.Other_error ("Expected explicit type", Parse_tree_sitter_helpers.token env _tok))
   | `Array_type x -> array_type env x
   | `Name x ->
       let n = name env x in

--- a/semgrep-core/src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml
+++ b/semgrep-core/src/parsing/tree_sitter/Parse_csharp_tree_sitter.ml
@@ -2308,7 +2308,9 @@ and type_ (env : env) (x : CST.type_) : G.type_ =
   | `Impl_type _tok ->
       (* When type_ is called, we expect an explicit type, not "var".
          The implicit type is handled in local_variable_type. *)
-      raise (Parse_info.Other_error ("Expected explicit type", Parse_tree_sitter_helpers.token env _tok))
+      raise
+        (Parse_info.Other_error
+           ("Expected explicit type", Parse_tree_sitter_helpers.token env _tok))
   | `Array_type x -> array_type env x
   | `Name x ->
       let n = name env x in

